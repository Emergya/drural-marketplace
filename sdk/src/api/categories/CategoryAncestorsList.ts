import BaseList, {
  GetPageInfo,
  GetTotalCount,
  MapQueryData,
} from "../../helpers/BaseList";
import {
  CategoryAncestorsList as CategoryAncestorsListQuery,
  CategoryAncestorsListVariables,
} from "../../queries/gqlTypes/CategoryAncestorsList";
import { BaseCategory } from "../../fragments/gqlTypes/BaseCategory";
import { categoryAncestorsList } from "../../queries/category";

export class CategoryAncestorsList extends BaseList<
  CategoryAncestorsListQuery,
  BaseCategory,
  CategoryAncestorsListVariables
> {
  getPageInfo: GetPageInfo<CategoryAncestorsListQuery> = result =>
    result.data.category?.ancestors?.pageInfo!;

  getTotalCount: GetTotalCount<CategoryAncestorsListQuery> = result =>
    result.data.category?.ancestors?.totalCount!;

  mapQueryData: MapQueryData<CategoryAncestorsListQuery, BaseCategory> = data =>
    data.category?.ancestors?.edges.map(({ node }) => node);

  query = (variables: CategoryAncestorsListVariables) =>
    this.client!.query<
      CategoryAncestorsListQuery,
      CategoryAncestorsListVariables
    >({
      fetchPolicy: "network-only",
      query: categoryAncestorsList,
      variables,
    });
}
