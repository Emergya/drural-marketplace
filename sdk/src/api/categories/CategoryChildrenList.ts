import BaseList, {
  GetPageInfo,
  GetTotalCount,
  MapQueryData,
} from "../../helpers/BaseList";
import {
  CategoryChildrenList as CategoryChildrenListQuery,
  CategoryChildrenListVariables,
} from "../../queries/gqlTypes/CategoryChildrenList";
import { BaseCategory } from "../../fragments/gqlTypes/BaseCategory";
import { categoryChildrenList } from "../../queries/category";

export class CategoryChildrenList extends BaseList<
  CategoryChildrenListQuery,
  BaseCategory,
  CategoryChildrenListVariables
> {
  getPageInfo: GetPageInfo<CategoryChildrenListQuery> = result =>
    result.data.category?.children?.pageInfo!;

  getTotalCount: GetTotalCount<CategoryChildrenListQuery> = result =>
    result.data.category?.children?.totalCount!;

  mapQueryData: MapQueryData<CategoryChildrenListQuery, BaseCategory> = data =>
    data.category?.children?.edges.map(({ node }) => node);

  query = (variables: CategoryChildrenListVariables) =>
    this.client!.query<
      CategoryChildrenListQuery,
      CategoryChildrenListVariables
    >({
      fetchPolicy: "network-only",
      query: categoryChildrenList,
      variables,
    });
}
