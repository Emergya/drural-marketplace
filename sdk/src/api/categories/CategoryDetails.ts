import BaseDetails, { MapQueryData } from "../../helpers/BaseDetails";
import {
  CategoryDetails as CategoryDetailsQuery,
  CategoryDetailsVariables,
} from "../../queries/gqlTypes/CategoryDetails";
import { CategoryDetails as Category } from "../../fragments/gqlTypes/CategoryDetails";
import { categoryDetails } from "../../queries/category";

export class CategoryDetails extends BaseDetails<
  CategoryDetailsQuery,
  Category,
  CategoryDetailsVariables
> {
  mapQueryData: MapQueryData<CategoryDetailsQuery, Category> = data =>
    data.category || undefined;

  query = (variables: CategoryDetailsVariables) =>
    this.client!.query<CategoryDetailsQuery, CategoryDetailsVariables>({
      fetchPolicy: "network-only",
      query: categoryDetails,
      variables,
    });
}
