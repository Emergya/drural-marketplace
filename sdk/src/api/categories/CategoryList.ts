import BaseList, {
  GetPageInfo,
  GetTotalCount,
  MapQueryData,
} from "../../helpers/BaseList";
import {
  CategoryList as CategoryListQuery,
  CategoryListVariables,
} from "../../queries/gqlTypes/CategoryList";
import { BaseCategory } from "../../fragments/gqlTypes/BaseCategory";
import { categoryList } from "../../queries/category";

export class CategoryList extends BaseList<
  CategoryListQuery,
  BaseCategory,
  CategoryListVariables
> {
  getPageInfo: GetPageInfo<CategoryListQuery> = result =>
    result.data.categories?.pageInfo!;

  getTotalCount: GetTotalCount<CategoryListQuery> = result =>
    result.data.categories?.totalCount!;

  mapQueryData: MapQueryData<CategoryListQuery, BaseCategory> = data =>
    data.categories?.edges.map(({ node }) => node);

  query = (variables: CategoryListVariables) =>
    this.client!.query<CategoryListQuery, CategoryListVariables>({
      fetchPolicy: "network-only",
      query: categoryList,
      variables,
    });
}
