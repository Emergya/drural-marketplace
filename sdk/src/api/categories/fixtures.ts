export const categoryWithChildren = "Q2F0ZWdvcnk6Nw==";
export const categoryWithParent = "Q2F0ZWdvcnk6MjI=";
