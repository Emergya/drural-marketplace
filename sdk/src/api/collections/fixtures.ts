export const collectionId = "Q29sbGVjdGlvbjox";
export const collectionSlug = "summer-collection";
