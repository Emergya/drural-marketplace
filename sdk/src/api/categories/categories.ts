import ApolloClient from "apollo-client";
import { CategoryAncestorsListVariables } from "../../queries/gqlTypes/CategoryAncestorsList";
import { CategoryChildrenListVariables } from "../../queries/gqlTypes/CategoryChildrenList";
import {
  CategoryDetails as CategoryDetailsQuery,
  CategoryDetailsVariables,
} from "../../queries/gqlTypes/CategoryDetails";
import { CategoryDetails as CategoryDetailsFragment } from "../../fragments/gqlTypes/CategoryDetails";
import {
  CategoryList as CategoryListQuery,
  CategoryListVariables,
} from "../../queries/gqlTypes/CategoryList";
import { BaseCategory } from "../../fragments/gqlTypes/BaseCategory";
import { WithDetails, WithList } from "../types";
import { CategoryList } from "./CategoryList";
import { CategoryAncestorsList } from "./CategoryAncestorsList";
import { CategoryChildrenList } from "./CategoryChildrenList";
import { CategoryDetails } from "./CategoryDetails";

export class CategoriesAPI
  implements
    WithDetails<
      CategoryDetailsQuery,
      CategoryDetailsFragment,
      CategoryDetailsVariables
    >,
    WithList<CategoryListQuery, BaseCategory, CategoryListVariables> {
  private client: ApolloClient<any>;

  constructor(client: ApolloClient<any>) {
    this.client = client;
  }

  /**
   * Method returning category details
   * @param variables Details parameters
   */
  getDetails = async (variables: CategoryDetailsVariables) => {
    const details = new CategoryDetails(this.client);

    await details.init(variables);

    return details;
  };

  /**
   * Method returning list of categories with ability to request next page
   * @param params List parameters
   */
  getList = async (variables: CategoryListVariables) => {
    const list = new CategoryList(this.client);

    await list.init(variables);

    return list;
  };

  getAncestors = async (variables: CategoryAncestorsListVariables) => {
    const list = new CategoryAncestorsList(this.client);

    await list.init(variables);

    return list;
  };

  getChildren = async (variables: CategoryChildrenListVariables) => {
    const list = new CategoryChildrenList(this.client);

    await list.init(variables);

    return list;
  };
}
