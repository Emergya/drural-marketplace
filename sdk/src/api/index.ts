import ApolloClient from "apollo-client";

import { LocalStorageManager } from "../data";
import { ApolloClientManager } from "../data/ApolloClientManager";
import { LocalStorageHandler } from "../helpers/LocalStorageHandler";
import { JobsManager } from "../jobs";
import { SaleorState } from "../state";
import { Config } from "../types";
import { AuthAPI } from "./Auth";
import APIProxy from "./APIProxy";
import { SaleorCartAPI } from "./Cart";
import { SaleorCheckoutAPI } from "./Checkout";
import { CollectionsAPI } from "./collections/collections";
import { CategoriesAPI } from "./categories/categories";
import { ProductsAPI } from "./products/products";

export * from "./Checkout";
export * from "./Cart";

export class SaleorAPI {
  auth: AuthAPI;

  checkout: SaleorCheckoutAPI;

  cart: SaleorCartAPI;

  categories: CategoriesAPI;

  collections: CollectionsAPI;

  products: ProductsAPI;

  /**
   * @deprecated Please do not use it anymore. Reference to API Proxy will be removed in future.
   * Now it just exists for legacy React hooks, which also will be removed.
   */
  legacyAPIProxy: APIProxy;

  constructor(
    client: ApolloClient<any>,
    apiProxy: APIProxy,
    config: Config,
    onStateUpdate?: () => any
  ) {
    this.legacyAPIProxy = apiProxy;

    const localStorageHandler = new LocalStorageHandler();
    const apolloClientManager = new ApolloClientManager(client);
    const jobsManager = new JobsManager(
      localStorageHandler,
      apolloClientManager
    );
    const saleorState = new SaleorState(
      config,
      localStorageHandler,
      apolloClientManager,
      jobsManager
    );
    const localStorageManager = new LocalStorageManager(
      localStorageHandler,
      saleorState
    );

    if (onStateUpdate) {
      saleorState.subscribeToNotifiedChanges(onStateUpdate);
    }

    this.auth = new AuthAPI(saleorState, jobsManager, config);
    this.checkout = new SaleorCheckoutAPI(saleorState, jobsManager, config);
    this.cart = new SaleorCartAPI(
      localStorageManager,
      apolloClientManager,
      saleorState,
      jobsManager,
      config
    );
    this.categories = new CategoriesAPI(client);
    this.collections = new CollectionsAPI(client, config);
    this.products = new ProductsAPI(client, config);

    this.legacyAPIProxy.attachAuthListener(authenticated => {
      if (!authenticated) {
        localStorageHandler.setCheckout({});
        localStorageHandler.setPayment({});
        localStorageHandler.setJobs(null);
      }
    });
  }
}
