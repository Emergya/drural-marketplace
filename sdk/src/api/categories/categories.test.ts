import ApolloClient from "apollo-client";
import { setupRecording, setupAPI } from "../../../testUtils/api";
import { CategoriesAPI } from "./categories";
import * as fixtures from "./fixtures";

setupRecording();

describe("Categories object", () => {
  let client: ApolloClient<any>;
  let categoriesAPI: CategoriesAPI;

  beforeAll(async done => {
    client = (await setupAPI()).client;
    categoriesAPI = new CategoriesAPI(client);

    done();
  });

  it("can get a details of category", async () => {
    const details = await categoriesAPI.getDetails({
      id: fixtures.categoryWithChildren,
    });

    expect(details.data).toMatchSnapshot();
    expect(details.loading).toBe(false);
  });

  it("can get a list of categories", async () => {
    const list = await categoriesAPI.getList({
      first: 20,
    });

    expect(list.data).toMatchSnapshot();
    expect(list.loading).toBe(false);
  });

  it("can get new page", async () => {
    const list = await categoriesAPI.getList({
      first: 1,
    });

    expect(list.data).toMatchSnapshot();
    expect(list.loading).toBe(false);

    list.next();

    expect(list.loading).toBe(true);

    await list.current;

    expect(list.data).toMatchSnapshot();
    expect(list.loading).toBe(false);
  });

  it("can get a list of subcategories", async () => {
    const list = await categoriesAPI.getChildren({
      first: 20,
      id: fixtures.categoryWithChildren,
    });

    expect(list.data).toMatchSnapshot();
    expect(list.loading).toBe(false);
  });

  it("can get a list of ancestor categories", async () => {
    const list = await categoriesAPI.getAncestors({
      first: 20,
      id: fixtures.categoryWithParent,
    });

    expect(list.data).toMatchSnapshot();
    expect(list.loading).toBe(false);
  });
});
