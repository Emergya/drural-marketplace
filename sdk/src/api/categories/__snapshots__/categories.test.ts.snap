// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Categories object can get a details of category 1`] = `
Object {
  "__typename": "Category",
  "backgroundImage": Object {
    "__typename": "Image",
    "alt": "",
    "url": "https://master.staging.saleor.cloud/media/category-backgrounds/accessories.jpg",
  },
  "description": "{\\"blocks\\": [{\\"data\\": {\\"text\\": \\"\\"}, \\"type\\": \\"paragraph\\"}]}",
  "id": "Q2F0ZWdvcnk6Nw==",
  "name": "Accessories",
  "seoDescription": "",
  "seoTitle": "",
  "slug": "accessories",
}
`;

exports[`Categories object can get a list of ancestor categories 1`] = `
Array [
  Object {
    "__typename": "Category",
    "id": "Q2F0ZWdvcnk6Nw==",
    "name": "Accessories",
    "seoDescription": "",
    "seoTitle": "",
    "slug": "accessories",
  },
]
`;

exports[`Categories object can get a list of categories 1`] = `
Array [
  Object {
    "__typename": "Category",
    "id": "Q2F0ZWdvcnk6Nw==",
    "name": "Accessories",
    "seoDescription": "",
    "seoTitle": "",
    "slug": "accessories",
  },
  Object {
    "__typename": "Category",
    "id": "Q2F0ZWdvcnk6OA==",
    "name": "Groceries",
    "seoDescription": "",
    "seoTitle": "",
    "slug": "groceries",
  },
  Object {
    "__typename": "Category",
    "id": "Q2F0ZWdvcnk6OQ==",
    "name": "Apparel",
    "seoDescription": "",
    "seoTitle": "",
    "slug": "apparel",
  },
  Object {
    "__typename": "Category",
    "id": "Q2F0ZWdvcnk6MTA=",
    "name": "T-shirts",
    "seoDescription": "",
    "seoTitle": "",
    "slug": "t-shirts",
  },
  Object {
    "__typename": "Category",
    "id": "Q2F0ZWdvcnk6MTE=",
    "name": "Polo Shirts",
    "seoDescription": "",
    "seoTitle": "",
    "slug": "polo-shirts",
  },
  Object {
    "__typename": "Category",
    "id": "Q2F0ZWdvcnk6MTI=",
    "name": "Hoodies",
    "seoDescription": "",
    "seoTitle": "",
    "slug": "hoodies",
  },
  Object {
    "__typename": "Category",
    "id": "Q2F0ZWdvcnk6MTM=",
    "name": "Footwear",
    "seoDescription": "",
    "seoTitle": "",
    "slug": "footwear",
  },
  Object {
    "__typename": "Category",
    "id": "Q2F0ZWdvcnk6MTQ=",
    "name": "Juices",
    "seoDescription": "",
    "seoTitle": "",
    "slug": "juices",
  },
  Object {
    "__typename": "Category",
    "id": "Q2F0ZWdvcnk6MTU=",
    "name": "Alcohol",
    "seoDescription": "",
    "seoTitle": "",
    "slug": "alcohol",
  },
  Object {
    "__typename": "Category",
    "id": "Q2F0ZWdvcnk6MjA=",
    "name": "Paints",
    "seoDescription": "",
    "seoTitle": "",
    "slug": "paints",
  },
  Object {
    "__typename": "Category",
    "id": "Q2F0ZWdvcnk6MjI=",
    "name": "Homewares",
    "seoDescription": "",
    "seoTitle": "",
    "slug": "homewares",
  },
  Object {
    "__typename": "Category",
    "id": "Q2F0ZWdvcnk6MjM=",
    "name": "Audiobooks",
    "seoDescription": "",
    "seoTitle": "",
    "slug": "ebooks",
  },
]
`;

exports[`Categories object can get a list of subcategories 1`] = `
Array [
  Object {
    "__typename": "Category",
    "id": "Q2F0ZWdvcnk6MjA=",
    "name": "Paints",
    "seoDescription": "",
    "seoTitle": "",
    "slug": "paints",
  },
  Object {
    "__typename": "Category",
    "id": "Q2F0ZWdvcnk6MjI=",
    "name": "Homewares",
    "seoDescription": "",
    "seoTitle": "",
    "slug": "homewares",
  },
  Object {
    "__typename": "Category",
    "id": "Q2F0ZWdvcnk6MjM=",
    "name": "Audiobooks",
    "seoDescription": "",
    "seoTitle": "",
    "slug": "ebooks",
  },
]
`;

exports[`Categories object can get new page 1`] = `
Array [
  Object {
    "__typename": "Category",
    "id": "Q2F0ZWdvcnk6Nw==",
    "name": "Accessories",
    "seoDescription": "",
    "seoTitle": "",
    "slug": "accessories",
  },
]
`;

exports[`Categories object can get new page 2`] = `
Array [
  Object {
    "__typename": "Category",
    "id": "Q2F0ZWdvcnk6Nw==",
    "name": "Accessories",
    "seoDescription": "",
    "seoTitle": "",
    "slug": "accessories",
  },
  Object {
    "__typename": "Category",
    "id": "Q2F0ZWdvcnk6OA==",
    "name": "Groceries",
    "seoDescription": "",
    "seoTitle": "",
    "slug": "groceries",
  },
]
`;
